server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoHashCellIdSource.java

```diff
@@ -73,50 +72,25 @@ protected int advanceValue(org.elasticsearch.common.geo.GeoPoint target, int val
 
     private static class BoundedCellValues extends CellValues {
 
+        private final GeoHashBoundedPredicate predicate;
         private final GeoBoundingBox bbox;
-        private final boolean crossesDateline;
 
         BoundedCellValues(MultiGeoPointValues geoValues, int precision, GeoBoundingBox bbox) {
             super(geoValues, precision);
+            this.predicate = new GeoHashBoundedPredicate(precision, bbox);
             this.bbox = bbox;
-            this.crossesDateline = bbox.right() < bbox.left();
         }
 
         @Override
         protected int advanceValue(org.elasticsearch.common.geo.GeoPoint target, int valuesIdx) {
             final String hash = Geohash.stringEncode(target.getLon(), target.getLat(), precision);
-            if (validPoint(target.getLon(), target.getLat()) || validHash(hash)) {
+            if (validPoint(target.getLon(), target.getLat()) || predicate.validHash(hash)) {
                 values[valuesIdx] = Geohash.longEncode(hash);
                 return valuesIdx + 1;
             }
             return valuesIdx;
         }
 
-        private boolean validHash(String hash) {
-            final Rectangle rect = Geohash.toBoundingBox(hash);
-            // hashes should not cross in theory the dateline but due to precision
-            // errors and normalization computing the hash, it might happen that they actually
-            // crosses the dateline.
-            if (rect.getMaxX() < rect.getMinX()) {
-                return intersects(-180, rect.getMaxX(), rect.getMinY(), rect.getMaxY())
-                    || intersects(rect.getMinX(), 180, rect.getMinY(), rect.getMaxY());
-            } else {
-                return intersects(rect.getMinX(), rect.getMaxX(), rect.getMinY(), rect.getMaxY());
-            }
-        }
-
-        private boolean intersects(double minX, double maxX, double minY, double maxY) {
-            // touching hashes are excluded
-            if (bbox.top() > minY && bbox.bottom() < maxY) {
-                if (crossesDateline) {
-                    return bbox.left() < maxX || bbox.right() > minX;
-                } else {
-                    return bbox.left() < maxX && bbox.right() > minX;
-                }
-            }
-            return false;
-        }
-
         private boolean validPoint(double x, double y) {
```

Comment:
It feels like this could be moved to `GeoHashBoundedPredicate`, just as the similar method from `GeoTileCellIdSource` was moved.

Reply:
I prefer not because `GeoHashBoundedPredicate` should only work on hash. This method is just an optimisation oof this class. If the function return true then we know it is a vlid hash, else we will still need to use the predicate.